t1rocketemu/src/AXI4SlaveAgent.scala

```diff
@@ -58,189 +73,156 @@ class ReadPayload(length: Int, dataWidth: Int) extends Bundle {
 class AXI4SlaveAgent(parameter: AXI4SlaveAgentParameter)
     extends FixedIORawModule[AXI4SlaveAgentInterface](new AXI4SlaveAgentInterface(parameter)) {
   dontTouch(io)
+  val invClock = (~io.clock.asBool).asClock;
+
   io.channel match {
-    case channel: AXI4RWIrrevocableVerilog =>
-      new WriteManager(channel)
-      new ReadManager(channel)
-    case channel: AXI4ROIrrevocableVerilog =>
-      new ReadManager(channel)
-    case channel: AXI4WOIrrevocableVerilog =>
-      new WriteManager(channel)
+    case channel: AXI4RWIrrevocableVerilog => {
+      val view = channel.viewAs[AXI4RWIrrevocable](
+        implicitly,
+        AXI4RWIrrevocableVerilog.viewChisel,
+        implicitly
+      )
+      new ReadManager(view)
+      new WriteManager(view)
+    }
+    case channel: AXI4ROIrrevocableVerilog => {
+      val view = channel.viewAs[AXI4ROIrrevocable](
+        implicitly,
+        AXI4ROIrrevocableVerilog.viewChisel,
+        implicitly
+      )
+      new ReadManager(view)
+    }
+    case channel: AXI4WOIrrevocableVerilog => {
+      val view = channel.viewAs[AXI4WOIrrevocable](
+        implicitly,
+        AXI4WOIrrevocableVerilog.viewChisel,
+        implicitly
+      )
+      new WriteManager(view)
+    }
+  }
+
+  RawClockedVoidFunctionCall(s"axi_tick")(
+    io.clock,
+    true.B,
+    io.reset.asTypeOf(UInt(8.W))
+  )
+
+  /// Widen a wire to make all DPI calls' types uniform
+  def widen(wire: UInt, target: Int): UInt = {
+    require(wire.isWidthKnown)
+    require(wire.getWidth <= target)
+    wire.asTypeOf(UInt(target.W))
   }
```

Comment:
It makes me nervous...
But anyway I this is OK....
please give a tag to align the axi channel id between RTL and ID.

Reply:
The tag is data_width which is given as another parameter in the DPI call.

The main problem is that since now we're using the same DPI function across all channels, it has to have a uniform type. Although the DPI side doesn't care about the width of the data as long as it's wider than 64 bit and is hence presented as a pointer, but the RTL side does.

If in the future, RTL code changes cause the channel to widen and exceed the width (`target`) given here, the assertion will fail, and the target width should be raised manually.